LibraBFT/Abstract/BFT.agda

```diff
@@ -228,6 +240,122 @@ module LibraBFT.Abstract.BFT
+   h∉t {_} {x₁ ∷ xs} {x} (on-∷ x< ∷ sxs) (here refl) = ⊥-elim (<⇒≢ x< refl)
+   h∉t {_} {x₁ ∷ xs} {x} (on-∷ x< ∷ (x₁< ∷ sxs)) (there x∈xs)
+     = h∉t ((trans-OnHead x₁< x<) ∷ sxs) x∈xs
+
+   y∉⇒All≢ : ∀ {n} {xs : List (Fin n)} {x y} → y ∉ (x ∷ xs)
+           → x ≢ y × y ∉ xs
+   y∉⇒All≢ {_} {xs} {x} {y} y∉
+     with y ∈? xs
+   ...| yes y∈xs = ⊥-elim (y∉ (there y∈xs))
+   ...| no  y∉xs
+     with x ≟Fin y
+   ...| yes x≡y = ⊥-elim (y∉ (here (sym x≡y)))
+   ...| no  x≢y = x≢y , y∉xs
+
+
+   ≤-head : ∀ {n} {xs : List (Fin n)} {x y}
+          → y ∈ (x ∷ xs) → IsSorted _<Fin_ (x ∷ xs)
+          → x ≤Fin y
+   ≤-head {_} {xs} {x} {x} (here refl) sxs = ≤-refl
+   ≤-head {_} {x₁ ∷ []} {x} {x₁} (there (here refl)) (on-∷ x< ∷ sxs) = <⇒≤ x<
+   ≤-head {_} {x₁ ∷ x₂ ∷ xs} {x} {y} (there y∈) (on-∷ x<x₁ ∷ sxs)
+     = ≤-trans (<⇒≤ x<x₁) (≤-head y∈ sxs)
+
+
+
+   _⊆List_ : ∀ {A : Set} → List A → List A → Set
+   xs ⊆List ys = All (_∈ ys) xs
+
+
+   ∈List-elim : ∀ {A : Set} {x y : A} {ys : List A}
+              → x ∈ (y ∷ ys) → x ≢ y
+              → x ∈ ys
+   ∈List-elim {_} {x} {y} {ys} (here x≡y) x≢y = ⊥-elim (x≢y x≡y)
+   ∈List-elim {_} {x} {y} {ys} (there x∈) x≢y = x∈
+
+
+   ⊆List-elim : ∀ {n} {y} {xs ys : List (Fin n)}
+              → xs ⊆List (y ∷ ys) → y ∉ xs
+              → xs ⊆List ys
+   ⊆List-elim {n} {y} {[]} {ys} xs∈ y∉ = []
+   ⊆List-elim {n} {x} {x ∷ xs} {[]} (here refl ∷ xs∈) y∉
+     = ⊥-elim (proj₁ (y∉⇒All≢ y∉) refl)
+   ⊆List-elim {n} {y} {x ∷ xs} {y₁ ∷ ys} (x∈ ∷ xs∈) y∉
+     with y∉⇒All≢ y∉
+   ...| x≢y , y∉xs = ∈List-elim x∈ x≢y ∷ ⊆List-elim xs∈ y∉xs
+
+
+   sort→∈-disj : ∀ {n} {x y} {xs ys : List (Fin n)}
+               → IsSorted _<Fin_ (x ∷ xs) → IsSorted _<Fin_ (y ∷ ys)
+               → x ∈ ys
+               → y ∉ xs
+   sort→∈-disj {ys = y₁ ∷ ys} (x< ∷ sxs) (on-∷ y<y₁ ∷ sys) x∈ys y∈xs
+     = let y₁≤x = ≤-head x∈ys sys
+           y<x  = <-transˡ y<y₁ y₁≤x
+       in h∉t (trans-OnHead x< y<x ∷ sxs) y∈xs
+
+
+   sum-⊆-≤ : ∀ {xs ys : List Member} (f : Member → ℕ)
+           → IsSorted _<Fin_ xs → IsSorted _<Fin_ ys
+           → xs ⊆List ys
+           → sum (List-map f xs) ≤ sum (List-map f ys)
+   sum-⊆-≤ {[]} {ys} f sxs sys [] = z≤n
+   sum-⊆-≤ {x ∷ xs} {x ∷ ys} f (x₁ ∷ sxs) (y₁ ∷ sys) (here refl ∷ xs∈)
+     = let xs∈ys = ⊆List-elim xs∈ (h∉t (x₁ ∷ sxs))
+       in +-monoʳ-≤ (f x) (sum-⊆-≤ f sxs sys xs∈ys)
+   sum-⊆-≤ {x ∷ xs} {y ∷ ys} f (x₁ ∷ sxs) (y₁ ∷ sys) (there px ∷ xs∈)
+     = let y∉xs  = sort→∈-disj (x₁ ∷ sxs) (y₁ ∷ sys) px
+           xs∈ys = ⊆List-elim xs∈ y∉xs
+       in ≤-stepsˡ (f y) (sum-⊆-≤ f (x₁ ∷ sxs) sys (px ∷ xs∈ys))
+
+
+   map-suc-sort : ∀ {n} {xs : List (Fin n)}
+                → IsSorted _<Fin_ xs
+                → IsSorted _<Fin_ (List-map suc xs)
+   map-suc-sort [] = []
+   map-suc-sort (x ∷ []) = [] ∷ []
+   map-suc-sort (on-∷ x< ∷ (x₁ ∷ sxs)) = (on-∷ (s≤s x<)) ∷ (map-suc-sort (x₁ ∷ sxs))
+
+
+   map-tabulate : ∀  {A B : Set} (n : ℕ) (g : A → B) (f : Fin n → A)
```

Comment:
This is in the library (`Data.List.Properties`) with the same name.

Reply:
Let's not talk about how many time it took me to prove this lemma :roll_eyes: